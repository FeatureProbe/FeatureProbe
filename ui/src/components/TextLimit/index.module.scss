.limit-str-container-w {
  display: inline-block;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  vertical-align: bottom;
  line-height: unset !important;
}
