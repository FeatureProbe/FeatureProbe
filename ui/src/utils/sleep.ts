const sleep = (delay: number) => {
  return new Promise((res) => {
    setTimeout(() => {
      res('ok');
    }, delay);
  });
};

export default sleep;
