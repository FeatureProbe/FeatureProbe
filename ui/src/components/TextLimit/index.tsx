import { ReactNode, useCallback, useEffect, useRef, useState } from 'react';
import { Popup, PopupProps } from 'semantic-ui-react';
import { stringLimit } from '../../utils/tools';
import styles from './index.module.scss';

interface IProps {
  text: string;
  maxLength?: number;
  hidePopup?: boolean;
  popupRender?: ReactNode;
  maxWidth?: number;
  popupProps?: PopupProps;
}

const TextLimit: React.FC<IProps> = (props) => {
  const { text, maxLength, maxWidth, hidePopup, popupRender, popupProps } = props;
  const ref = useRef<HTMLDivElement | null>(null);
  const [isLong, setIsLong] = useState<boolean>(false);

  const judgeLength: () => boolean = useCallback(() => {
    if (maxLength) {
      return text.length > maxLength;
    } else {
      if (ref.current) {
        return ref.current.scrollWidth > ref.current.clientWidth;
      } else {
        return false;
      }
    }
  }, [text, maxLength]);

  useEffect(() => {
    setIsLong(judgeLength());
  }, [judgeLength, text]);

  return hidePopup ? (
    <span
      className={`${maxLength ? styles['limit-str-container-n'] : styles['limit-str-container-w']}`}
      style={{
        maxWidth: maxWidth ?? '100%',
      }}
    >
      {stringLimit(text, maxLength ?? 0)}
    </span>
  ) : (
    <Popup
      inverted
      disabled={!isLong}
      trigger={
        <span
          className={`${maxLength ? styles['limit-str-container-n'] : styles['limit-str-container-w']}`}
          ref={ref}
          style={{
            maxWidth: maxWidth ?? '100%',
          }}
        >
          {stringLimit(text, maxLength ?? 0)}
        </span>
      }
      content={popupRender ?? text}
      className="popup-override"
      wide="very"
      {...popupProps}
    />
  );
};

export default TextLimit;
