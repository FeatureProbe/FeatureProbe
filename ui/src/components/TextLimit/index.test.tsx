import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import TextLimit from '.';
import sleep from 'utils/sleep';

it('TextLimit snapshot', () => {
  (async () => {
    const { asFragment, rerender } = render(<TextLimit text="These are test words." maxLength={10} />);
    expect(asFragment()).toMatchSnapshot();

    rerender(<TextLimit text="These are test words." maxWidth={40} />);
    expect(asFragment()).toMatchSnapshot();

    rerender(
      <TextLimit popupRender={<div>test</div>} text="These are test words." maxLength={1000} />
    );
    expect(asFragment()).toMatchSnapshot();

    rerender(
      <TextLimit hidePopup text="These are test words." maxWidth={1000} />
    );
    expect(asFragment()).toMatchSnapshot();
  })();
});

console.error = jest.fn;

test('TextLimit text maxLenth test', (done) => {
  (async () => {
    const { baseElement } = render(<TextLimit text="This is a test text." maxLength={10} />);

    expect(screen.getByText('This is...')).toBeInTheDocument();
    await userEvent.hover(screen.getByText('This is...'));
    await sleep(200);

    expect(baseElement).toMatchSnapshot();
    done();
  })();
});

test('TextLimit text maxWidth test', (done) => {
  (async () => {
    const { baseElement } = render(<TextLimit text="This is a test text." maxWidth={50} />);

    expect(screen.getByText('This is a test text.')).toBeInTheDocument();
    await userEvent.hover(screen.getByText('This is a test text.'));
    await sleep(200);

    expect(baseElement).toMatchSnapshot();
    done();
  })();
});
