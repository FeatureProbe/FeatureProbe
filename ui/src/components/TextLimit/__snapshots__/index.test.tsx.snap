// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextLimit snapshot 1`] = `
<DocumentFragment>
  <span
    class="limit-str-container-n"
    style="max-width: 100%;"
  >
    These a...
  </span>
</DocumentFragment>
`;

exports[`TextLimit snapshot 2`] = `
<DocumentFragment>
  <span
    class="limit-str-container-w"
    style="max-width: 40px;"
  >
    These are test words.
  </span>
</DocumentFragment>
`;

exports[`TextLimit snapshot 3`] = `
<DocumentFragment>
  <span
    class="limit-str-container-n"
    style="max-width: 100%;"
  >
    These are test words.
  </span>
</DocumentFragment>
`;

exports[`TextLimit snapshot 4`] = `
<DocumentFragment>
  <span
    class="limit-str-container-w"
    style="max-width: 1000px;"
  >
    These are test words.
  </span>
</DocumentFragment>
`;

exports[`TextLimit text maxLenth test 1`] = `
<body>
  <div>
    <span
      class="limit-str-container-n"
      style="max-width: 100%;"
    >
      This is...
    </span>
  </div>
  <div
    style="display: flex; position: absolute; left: 0px; top: 0px; bottom: 0px; transform: translate(0px, 0px);"
  >
    <div
      class="ui top left very wide inverted popup transition visible popup-override"
      style="position: initial;"
    >
      <div
        class="content"
      >
        This is a test text.
      </div>
    </div>
  </div>
</body>
`;

exports[`TextLimit text maxWidth test 1`] = `
<body>
  <div>
    <span
      class="limit-str-container-w"
      style="max-width: 50px;"
    >
      This is a test text.
    </span>
  </div>
</body>
`;
